lib/Service/ContactsIntegration.php

```diff
@@ -26,39 +26,74 @@
 
 use OCP\Contacts\IManager;
 use OCP\IConfig;
+use OCP\IGroupManager;
+use OCP\IUserManager;
 
 class ContactsIntegration {
 	/** @var IManager */
 	private $contactsManager;
 
+	/** @var IGroupManager */
+	private $groupManager;
+
+	/** @var IUserManager */
+	private $userManager;
+
 	/** @var IConfig */
 	private $config;
 
-	public function __construct(IManager $contactsManager, IConfig $config) {
+	public function __construct(IManager $contactsManager,
+								IGroupManager $groupManager,
+								IUserManager $userManager,
+								IConfig $config) {
 		$this->contactsManager = $contactsManager;
+		$this->groupManager = $groupManager;
+		$this->userManager = $userManager;
 		$this->config = $config;
 	}
 
 	/**
 	 * Extracts all matching contacts with email address and name
 	 *
+	 * @param string $userId
 	 * @param string $term
 	 * @return array
 	 */
-	public function getMatchingRecipient(string $term): array {
+	public function getMatchingRecipient(string $userId, string $term): array {
 		if (!$this->contactsManager->isEnabled()) {
 			return [];
 		}
 
 		// If 'Allow username autocompletion in share dialog' is disabled in the admin sharing settings, then we must not
 		// auto-complete system users
-		$allowSystemUsers = $this->config->getAppValue('core', 'shareapi_allow_share_dialog_user_enumeration', 'no') === 'yes';
+		$shareeEnumeration = $this->config->getAppValue('core', 'shareapi_allow_share_dialog_user_enumeration', 'no') === 'yes';
+		$shareeEnumerationInGroupOnly = $this->config->getAppValue('core', 'shareapi_restrict_user_enumeration_to_group', 'no') === 'yes';
+		$shareeEnumerationFullMatch = $this->config->getAppValue('core', 'shareapi_restrict_user_enumeration_full_match', 'yes') === 'yes';
+		$shareeEnumerationFullMatchUserId = $shareeEnumerationFullMatch && $this->config->getAppValue('core', 'shareapi_restrict_user_enumeration_full_match_userid', 'yes') === 'yes';
+		$shareeEnumerationFullMatchEmail = $shareeEnumerationFullMatch && $this->config->getAppValue('core', 'shareapi_restrict_user_enumeration_full_match_email', 'yes') === 'yes';
+		if ($shareeEnumeration && $shareeEnumerationInGroupOnly) {
```

Comment:
```suggestion
Move the contactsManager query before the `if` check
```suggestion
		$result = $this->contactsManager->search($term, ['UID', 'FN', 'EMAIL'], ['enumeration' => $shareeEnumeration]);
		if(empty($result)) {
			return [];
		}
		$receivers = [];

		if ($shareeEnumerationInGroupOnly) {
```
```